### ISSUE TITLE
Methods that clone a partially mocked parameter lead to infinite recursion when there is a missing expectation

### ISSUE THREAD

**Issue by alexweissman** [mockery/mockery#777]
I'm running PHP 5.6.15.

I have a [method](https://github.com/userfrosting/UserFrosting/blob/62f382d8ec0ab0cc62cffd6b7b9af2e87d20bd64/app/sprinkles/core/src/Database/Relations/Concerns/Unique.php#L257-L298) that, as its first step, clones its instance of an `Eloquent\Builder` parameter.

Since `Eloquent\Builder` has a member `Query\Builder` that it passes through many of its method calls to, I ended up having to do a partial mock:

```
$relation = $this->getRelation();
$baseQuery = m::mock('Illuminate\Database\Query\Builder');
$builder = m::mock('Illuminate\Database\Eloquent\Builder', [$baseQuery])->makePartial();

// Expectations on $builder
...

$paginatedQuery = $relation->getPaginatedQuery($builder, 2, 1);
```

When all of my expectations on `$builder` are correctly declared, the test passes without a problem.  However if I am missing an expectation, instead of getting the usual "Call to undefined method" error, PHPUnit halts with a `Segmentation fault 11` error.

It's a difficult problem to unpack - obviously the problem ultimately lies somewhere in PHP itself, but it's difficult to determine whether it is Mockery or Laravel that is triggering this problem.

**Comment by GrahamCampbell** [comment]
> lead to a segmentation fault

A segfault will always be a bug in PHP itself.

That said, a common cause of segfaults can be infinite loops. Try installing xdebug, and see if it's callstack limited halts your program.

**Comment by GrahamCampbell** [comment]
What happens on the newest php 5.6.x please?

**Comment by alexweissman** [comment]
I'll try both of these suggestions.  I don't have an easy way to switch between versions of PHP in my dev environment at the moment, so it might take a little while to get that set up.

**Comment by alexweissman** [comment]
Oh, xdebug is very useful!  Here's the truncated callstack:

```
Fatal error: Maximum function nesting level of '100' reached, aborting! in /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code on line 656

Call Stack:
    0.0008     236672   1. {main}() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/phpunit:0
    0.0107    1188392   2. PHPUnit_TextUI_Command::main() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/phpunit:52
    0.0108    1189144   3. PHPUnit_TextUI_Command->run() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/TextUI/Command.php:116
    0.0920    6778904   4. PHPUnit_TextUI_TestRunner->doRun() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/TextUI/Command.php:186
    0.1049    7023544   5. PHPUnit_Framework_TestSuite->run() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:517
    0.2986   16919784   6. PHPUnit_Framework_TestSuite->run() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestSuite.php:722
    0.3057   16923368   7. PHPUnit_Framework_TestCase->run() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestSuite.php:722
    0.3057   16923992   8. PHPUnit_Framework_TestResult->run() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:860
    0.3058   16926928   9. PHPUnit_Framework_TestCase->runBare() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestResult.php:686
    0.3119   17046304  10. PHPUnit_Framework_TestCase->runTest() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:905
    0.3119   17047152  11. ReflectionMethod->invokeArgs() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:1054
    0.3119   17047872  12. UserFrosting\Tests\Unit\BelongsToManyThroughTest->testPaginatedQuery() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:1054
    0.5704   22630736  13. UserFrosting\Sprinkle\Core\Database\Relations\BelongsToManyThrough->getPaginatedQuery() /Users/alexw/dev/htdocs/UserFrosting/app/sprinkles/core/tests/Unit/BelongsToManyThroughTest.php:68
    0.5719   22639336  14. Mockery_5_Illuminate_Database_Eloquent_Builder->whereIn() /Users/alexw/dev/htdocs/UserFrosting/app/sprinkles/core/src/Database/Relations/Concerns/Unique.php:296
    0.5719   22639776  15. Mockery_5_Illuminate_Database_Eloquent_Builder->__call() /Users/alexw/dev/htdocs/UserFrosting/app/sprinkles/core/src/Database/Relations/Concerns/Unique.php:296
    0.5719   22639776  16. Mockery_5_Illuminate_Database_Eloquent_Builder->_mockery_handleMethodCall() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:325
    0.5720   22640168  17. call_user_func_array:{/Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694}() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
    0.5720   22640760  18. Illuminate\Database\Eloquent\Builder->whereIn() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
    0.5720   22641200  19. Mockery_5_Illuminate_Database_Eloquent_Builder->__call() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:0
    0.5720   22641200  20. Mockery_5_Illuminate_Database_Eloquent_Builder->_mockery_handleMethodCall() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:325
    0.5722   22641592  21. call_user_func_array:{/Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694}() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
    0.5722   22642184  22. Illuminate\Database\Eloquent\Builder->whereIn() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
    0.5722   22642624  23. Mockery_5_Illuminate_Database_Eloquent_Builder->__call() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:0
    0.5722   22642624  24. Mockery_5_Illuminate_Database_Eloquent_Builder->_mockery_handleMethodCall() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:325
    0.5723   22643016  25. call_user_func_array:{/Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694}() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
    0.5723   22643608  26. Illuminate\Database\Eloquent\Builder->whereIn() /Users/alexw/dev/htdocs/UserFrosting/app/vendor/mockery/mockery/library/Mockery/Loader/EvalLoader.php(16) : eval()'d code:694
```

So it's clearly an infinite recursion.  As I suspected, it has something to do with `Eloquent\Builder` passing through method calls to `Query\Builder`.  However, I am not sure why this leads to an infinite loop!

**Comment by alexweissman** [comment]
Also, just realized I opened a lucky issue number!  Do I get a prize?  🏆

**Comment by alexweissman** [comment]
Confirmed the same behavior in PHP 5.6.31.

**Comment by alexweissman** [comment]
This appears to be the same issue as with #461.  However, I don't agree that I should need to cram all of my expectations into an array parameter in the constructor.

**Comment by alexweissman** [comment]
Ah!  It would appear that this behavior has been fixed as of 1.0.0-alpha.  I would definitely recommend mentioning this in the changelog :-D

Thanks for the help.  Seems like xdebug helped a lot in determining the root issue, which in turn led me to the relevant issues in the issue tracker.
